# CÂU 3: PHÂN TÍCH LÝ DO TRẢ HÀNG THEO DANH MỤC

**1. Yêu cầu bài toán:**
Thực hiện kết nối (join) giữa hai file `returns.csv` và `products.csv` theo `product_id` để tìm lý do trả hàng phổ biến nhất của danh mục sản phẩm **Streetwear**.

**2. Phương pháp tiếp cận & Suy luận:**
Để trả lời câu hỏi này, chúng ta cần thông tin từ cả 2 bảng dữ liệu, do đó quy trình sẽ gồm:
* **Bước 1:** Tải dữ liệu từ 2 file `returns.csv` (chứa lý do trả hàng) và `products.csv` (chứa thông tin danh mục).
* **Bước 2:** Kết nối (Merge/Join) hai bảng này lại với nhau dựa trên cột chung là `product_id`.
* **Bước 3:** Lọc dữ liệu chỉ lấy những dòng có danh mục (`category`) là 'Streetwear'.
* **Bước 4:** Đếm tần suất xuất hiện của từng lý do trả hàng (`return_reason`) trong tập dữ liệu vừa lọc và tìm ra lý do có số lượng cao nhất.

**3. Lý do chọn phương pháp:**
* **Sử dụng `pd.merge()`:** Đây là hàm mạnh mẽ nhất của Pandas để mô phỏng lại lệnh `JOIN` trong cơ sở dữ liệu quan hệ (SQL). Việc nối bảng bằng mã định danh (`product_id`) giúp đảm bảo tính toàn vẹn dữ liệu, không bị sai lệch thông tin khi ghép.
* **Sử dụng `value_counts()`:** Thay vì phải viết vòng lặp để đếm thủ công, `value_counts()` là hàm được tối ưu hóa bằng ngôn ngữ C bên dưới Pandas, giúp đếm tần suất các giá trị phân loại (categorical data) và tự động sắp xếp từ cao xuống thấp chỉ trong tích tắc.

In [1]:
# Import thư viện
import pandas as pd
from google.colab import drive

# Kết nối với Google Drive
drive.mount('/content/drive')

Mounted at /content/drive


**Thực hiện Bước 1 & Bước 2: Tải dữ liệu và Nối bảng (Merge)**

Ta sẽ đọc cả 2 file, sau đó dùng lệnh ghép bảng (inner join) để gắn thêm thông tin về 'category' vào từng dòng dữ liệu trả hàng tương ứng.

In [2]:
# 1. Khai báo đường dẫn
path_returns = '/content/drive/MyDrive/datathon-2026-round-1/returns.csv'
path_products = '/content/drive/MyDrive/datathon-2026-round-1/products.csv'

# 2. Đọc dữ liệu
df_returns = pd.read_csv(path_returns)
df_products = pd.read_csv(path_products)

# 3. Thực hiện Join 2 bảng theo product_id
# 'inner' join sẽ giữ lại các dòng có product_id tồn tại ở cả 2 bảng
df_merged = pd.merge(df_returns, df_products, on='product_id', how='inner')

# Hiển thị vài cột quan trọng của 5 dòng đầu để kiểm tra xem đã nối thành công chưa
df_merged[['return_id', 'product_id', 'return_reason', 'category']].head()

,return_id,product_id,return_reason,category
0,RET-000001,609,late_delivery,Streetwear
1,RET-000002,1862,wrong_size,GenZ
2,RET-000003,2359,wrong_size,Streetwear
3,RET-000004,1449,wrong_size,Outdoor
4,RET-000005,1450,wrong_size,Outdoor


**Thực hiện Bước 3 & Bước 4: Lọc dữ liệu và Thống kê**

Từ chiếc bảng khổng lồ vừa nối, ta chỉ "trích xuất" ra nhóm hàng 'Streetwear', sau đó dùng hàm đếm để tìm ra thủ phạm chính khiến khách hàng trả đồ.

In [3]:
# Lọc các bản ghi thuộc danh mục Streetwear
df_streetwear = df_merged[df_merged['category'] == 'Streetwear']

# Đếm tần suất từng lý do trả hàng
reason_counts = df_streetwear['return_reason'].value_counts()

print("Thống kê số lượng theo từng lý do trả hàng (Streetwear):")
print(reason_counts)

# Lấy lý do đứng đầu (phổ biến nhất)
top_reason = reason_counts.idxmax()
print(f"\n--> Lý do trả hàng phổ biến nhất là: '{top_reason}'")

Thống kê số lượng theo từng lý do trả hàng (Streetwear):
return_reason
wrong_size          7626
defective           4330
not_as_described    3854
changed_mind        3830
late_delivery       2159
Name: count, dtype: int64

--> Lý do trả hàng phổ biến nhất là: 'wrong_size'


**KẾT LUẬN CUỐI CÙNG:**
Dựa trên kết quả đếm tần suất của bảng dữ liệu đã nối, lý do trả hàng phổ biến nhất đối với danh mục Streetwear là **wrong_size** (với 7626 lượt trả hàng).

Đối chiếu với các phương án của đề bài:

A) defective

B) wrong_size

C) changed_mind

D) not_as_described

**=> Đáp án được chọn là: B**